multi-region: conditional bootstrap

bernadinm opened the pull request:
This adds control to enable a bootstrap node on the remote region

bernadinm committed:
adding enable_bootstrap variable

bernadinm committed:
adding temp remote branch for testing

fatz commented on main.tf:
```diff
@@ -61,9 +61,11 @@
+  #source  = "dcos-terraform/dcos-install-bootstrap-remote-exec/null"
+  #version = "~> 0.1"
 
+  enable_bootstrap     = "${var.enable_bootstrap}"
```
why do we need to be explicit on this? A simple length check on bootstrap_ip should do the same

bernadinm replied:
the reason is because when the remote region exist, there actually exist a bootstrap_ip, however it is for the local region. When I tested this, I assumed that it won't be needed to be explicitly called, however this caused the remote region to try to rerun the bootstrap process on the local region. This explicitly causes this to be turned off. 

Now this looks like this when adding `enable_bootstrap`. Notice how this is called in the remote region section below.

```
module "remote-1" {
  source = "dcos-terraform/remote-agents/aws"
  version = "~> 0.1.0"

  providers = {
    aws = "aws.remote-1"
  }
  dcos_instance_os     = "coreos_1855.5.0"
  cluster_name         = "region2-cluster2"
  ssh_public_key_file  = "~/.ssh/id_rsa.pub"
  dcos_version         = "1.12.1"
  admin_ips            = ["${data.http.whatismyip.body}/32"]

  # enable_bootstrap     = true
  master_private_ips     = "${module.dcos.infrastructure-masters.private_ips}"

  enable_bootstrap     = false
  bootstrap_private_ip = "${module.dcos.infrastructure-bootstrap.private_ip}"

  # review with SRE
  # bootstrap_prereq-id  = "${module.dcos.infrastructure-bootstrap.prereq-id}"
  # masters_prereq-id  = "${module.dcos.infrastructure-masters.prereq-id}"

  num_private_agents = "3"
  num_public_agents  = "1"
  subnet_range       = "172.13.0.0/16"

  dcos_variant              = "ee"
  dcos_license_key_contents = "${file("./license.txt")}"

  tags={owner = "mbernadin", expiration = "6h"}
  dcos_install_mode  = "${var.dcos_install_mode}"
}
```

fatz replied:
but why not having a condition on `bootstrap_private_ip`

if `bootstrap_private_ip` is set in that remote module... no need for spawning a bootstrap node. If its not set you need a local one and therefore spawn it and forward its public and private address to the install part

bernadinm replied:
This will stop the `null_resource` from running on the remote region when there doesn't exist a remote bootstrap node there but a local region bootstrap node exists. Without this I noticed that when only one bootstrap node exists and there are two resource `"null_resource" "bootstrap"`, it will try to perform two bootstrap ssh commands on the same machine. By enabling an `enable_bootstrap` flag, this tells the remote module to skip the first bootstrap node because its already handled by another core module and it doesn't need to configure it but it can be called and referenced by the private agents and public agents within the `dcos-install` layer.

bernadinm committed:
Merge branch 'master' into conditional-bootstrap

bernadinm committed:
fmt: formatting and readme

bernadinm commented:
This depends on https://github.com/dcos-terraform/terraform-null-dcos-install-bootstrap-remote-exec/pull/3

Commit:
Error: module "dcos-bootstrap-install": "enable_bootstrap" is not a valid argument
Getting the following error while trying to test this. Setting the source to appropriate branch that contains the variable. It looks like the source was moved back via commit ce642c9d8810a5832fb322be26f88782703a2ad8 (merge master into this branch). Not sure this was intentional or missed. Can change this back once we discuss current status on this.

bernadinm committed:
main: reverting the sources back to original

bernadinm committed:
Merge branch 'master' into conditional-bootstrap

bernadinm commented:
Updated the source back to the original

geekbass reviewed (approved):
LGTM. Ty!
